### ISSUE TITLE
Do you have a strong opinion about EventEmitter?

### ISSUE THREAD

**Issue by humanchimp** [kriskowal/ama#2]
I found [this comment of yours](https://github.com/nodejs/post-mortem/issues/16#issuecomment-181606242) to be interesting, but I wondered if you could offer a little more context.

**Comment by kriskowal** [comment]
This comment was a tongue-in-cheek reply to @raynos. We have worked together with @jcorbin at Uber on projects including [TChannel](https://github.com/uber/tchannel) and [Hyperbahn](https://github.com/uber/hyperbahn). Jake and Josh observed from iterating on performance improvements informed by production flame graphs that the EventEmitter was unnecessarily wasteful, and created an [alternative](https://github.com/uber/tchannel-node/blob/master/lib/event_emitter.js).

The alternative allows us to bypass a dictionary lookup from event name to emitter state, by placing the emitter state and interface on a property of the event emitter itself.

``` js
function Beeper() {
    this.beepEvent = this.defineEvent("beep");
}
util.inherits(Beeper, EventEmitter);

var beeper = new Beeper();
beeper.beepEvent.on(onBeep, recorder);
beeper.beepEvent.emit({volume: 1});

function onBeep(beep, recorder) {
}
```

Consequently, looking up an event benefits from v8 hidden classes. We also observed that we do not need an array in the common case of a single observer, reducing garbage collection. The new event emitter also sends an observer-selected context object, which helps us re-use functions for multiple consumers and avoid closures in many cases.

You will note that Node.js does not use event emitters internally. This optimization is just one of many steps in the walk away from abstractions to high performance. The next logical step is to completely eliminate the notion of extensible subscription and just have delegates or hooks for all of the anticipated consumers of an event and dispatch on them with individual, manually ordered method calls, much like DOM level 1 events like "onclick".

In short, there are trade-offs. I think that we all go home happy if we take an opt-in layered approach that includes both higher _and_ lower layers than what Node.js currently exposes. The higher level abstractions trade performance for robust composition, and the lower would trade ergonomics for speed.

**Comment by Raynos** [comment]
I have similar opinions that event emitter is not needed.

The only thing I like about it is an error handling contract. "Ive just allocated this thing that may error in the future please deal with it lol"

**Comment by humanchimp** [comment]
Thanks for the thorough reply!

**Comment by kriskowal** [comment]
Opinon refresh:

If you’re reaching for an EventEmitter, you would be better served with something else:

- If you want to call back for a kind of “event”, use a callback function. If you need a bank of callbacks, use a delegate object with named methods. If you’re in TypeScript, define the interface of the delegate. The ceremony that EventEmitter saves you is trivial to write inline, provides better type signatures, is easier to document, and will be faster. If the provider of the event feels they need to multiplex, they can provide a callback that broadcasts to other callbacks.

- If you want an object stream, use an async iterator. It’s the language’s lingua franca and supports back-pressure and all sorts of combinators in terms of ordinary loops. Note that the iterator protocol calls `return()` if a loop exits, so you can support single-subscriber and multi-subscriber disposal patterns.

**Comment by Raynos** [comment]
If multiple things need to wait for something to complete implement a few boilerplate methods on your class

```
class DoComplexShit {
  waiters: Array<() => void>

  complicatedImplementation() {
    /** ... */
    for (const p of this.waiters) p()
    /** ... */
  }

  waitForComplete(): Promise<void> {
    return new Promise((resolve) => {
      this.waiters.push(resolve)
    })
  }
}
```

Don't inherit EventEmitter and emit a complete event, don't litter `things.once('complete', ...)` upstream.
